simplefs/simplefs.go

```diff
@@ -485,82 +485,115 @@ func (k *SimpleFS) SimpleFSList(ctx context.Context, arg keybase1.SimpleFSListAr
+
+		fs, finalElem, err := k.getFS(ctx, path)
+		switch err.(type) {
+		case nil:
+		case libfs.TlfDoesNotExist:
+			// TLF doesn't exist yet; just return an empty result.
+			k.setResult(opID, keybase1.SimpleFSListResult{})
+			return nil
+		default:
+			return err
+		}
+
+		// With listing, we don't know the totals ahead of time,
+		// so just start with a 0 total.
+		k.setProgressTotals(opID, 0, 0)
+		fi, err := fs.Stat(finalElem)
+		if err != nil {
+			return err
+		}
+		var des []keybase1.Dirent
+		if !fi.IsDir() {
+			var d keybase1.Dirent
+			err := setStat(&d, fi)
+			if err != nil {
 				return err
 			}
+			des = append(des, d)
+			// Leave paths empty so we can skip the loop below.
+		} else {
+			// Start with a depth of 0
+			paths = append(paths, pathStackElem{finalElem, 0})
+		}
 
-			// With listing, we don't know the totals ahead of time,
-			// so just start with a 0 total.
-			k.setProgressTotals(arg.OpID, 0, 0)
-			fi, err := fs.Stat(finalElem)
+		for len(paths) > 0 {
+			// Take last element and shorten.
+			pathElem := paths[len(paths)-1]
+			paths = paths[:len(paths)-1]
+
+			fis, err := fs.ReadDir(pathElem.path)
 			if err != nil {
 				return err
 			}
-			var des []keybase1.Dirent
-			if !fi.IsDir() {
-				var d keybase1.Dirent
-				err := setStat(&d, fi)
-				if err != nil {
-					return err
+			for _, fi := range fis {
+				// We can only get here if we're listing a
+				// directory, not a single file, so we should
+				// always filter.
+				if isFiltered(filter, fi.Name()) {
+					continue
 				}
-				des = append(des, d)
-				// Leave paths empty so we can skip the loop below.
-			} else {
-				paths = append(paths, finalElem)
-			}
 
-			for len(paths) > 0 {
-				// Take last element and shorten.
-				path := paths[len(paths)-1]
-				paths = paths[:len(paths)-1]
-
-				fis, err := fs.ReadDir(path)
+				var de keybase1.Dirent
+				err := setStat(&de, fi)
 				if err != nil {
 					return err
 				}
-				for _, fi := range fis {
-					// We can only get here if we're listing a
-					// directory, not a single file, so we should
-					// always filter.
-					if isFiltered(arg.Filter, fi.Name()) {
-						continue
-					}
-
-					var de keybase1.Dirent
-					err := setStat(&de, fi)
-					if err != nil {
-						return err
-					}
-					des = append(des, de)
-					if fi.IsDir() {
-						paths = append(paths, stdpath.Join(path, fi.Name()))
-					}
+				des = append(des, de)
+				// Only recurse if the caller requested infinite depth (-1), or
+				// if the current path has a depth less than the desired final
+				// depth of recursion.
+				if fi.IsDir() && (finalDepth == -1 || pathElem.depth < finalDepth) {
+					paths = append(paths, pathStackElem{stdpath.Join(pathElem.path, fi.Name()), pathElem.depth + 1})
 				}
-				k.updateReadProgress(arg.OpID, 0, int64(len(fis)))
 			}
-			k.setResult(arg.OpID, keybase1.SimpleFSListResult{Entries: des})
+			k.updateReadProgress(opID, 0, int64(len(fis)))
```

Comment:
It kind of seems like this should only add one to the FilesRead progress, but should update the FilesTotal progress by len(fis). (And you'd start the whole thing out with a total of 1.) That way the progress tracking will be a little more useful as it can kinda predict the latency of the ReadDirs.

Maybe that's overkill though.

Reply:
Hmm. I'm not really sure what the implications are on the UX with respect to listing directories. I follow your argument, but not sure what the right answer is. @songgao WDYT?